src/components/Label/Label.tsx

```diff
@@ -41,3 +106,67 @@ function hslToHex(h: number, s: number, lightness: number) {
   };
   return `#${f(0)}${f(8)}${f(4)}`;
 }
+
+export function filterNotesByLabels(
+  labels: ILabel[],
+  { onlyInactive }: { onlyInactive: boolean } = { onlyInactive: true },
+): IStorageNote[] {
+  return filterDecoratedNotesByLabels(labels, { hasAllLabels: true, onlyInactive: onlyInactive }).map(
+    (note) => note.original,
+  );
+}
+
+export function filterDecoratedNotesByLabels(
+  labels: ILabel[],
+  { hasAllLabels, onlyInactive }: { hasAllLabels: boolean; onlyInactive: boolean } = {
+    hasAllLabels: true,
+    onlyInactive: false,
+  },
+): IDecoratedNote[] {
+  const notesSet = new Set<IDecoratedNote>();
+  const inactiveNotesSet = new Set<IDecoratedNote>();
+
+  function traverseAndCollectNotes(label: ILabel, notes: Set<IDecoratedNote>, isActive = true) {
+    if (label.isActive === isActive) {
+      for (const note of label.notes) {
+        notes.add(note);
+      }
+    }
+    if (label.children) {
+      for (const child of label.children) {
+        traverseAndCollectNotes(child, notes, isActive);
+      }
+    }
+  }
+
+  if (onlyInactive) {
+    for (const label of labels) {
+      traverseAndCollectNotes(label, inactiveNotesSet, false);
+    }
+    return Array.from(inactiveNotesSet);
+  }
+
+  for (const label of labels) {
+    traverseAndCollectNotes(label, notesSet);
+  }
+  if (hasAllLabels) {
+    for (const label of labels) {
+      traverseAndCollectNotes(label, inactiveNotesSet, false);
+    }
+    for (const note of inactiveNotesSet) {
+      notesSet.delete(note);
+    }
+  }
+
+  return Array.from(notesSet);
```

Comment:
```suggestion
  function traverseAndCollectNotes(labels: ILabel[], notes: Set<IDecoratedNote>, isActive = true) {
    for (const label of labels) {
      if (label.isActive === isActive) {
        for (const note of label.notes) {
          notes.add(note);
        }
      }
    
      if (label.children) {
        traverseAndCollectNotes(label.children, notes, isActive);
      }
    }
    return notes;
  }

  // return notes from inactive labels
  if (onlyInactive) {
    return Array.from(
      traverseAndCollectNotes(labels, new Set(), false)
    );
  }

  const activeNotes = traverseAndCollectNotes(labels, notesSet);
  if (hasAllLabels) {
    const inactiveNotes = traverseAndCollectNotes(labels, new Set(), false);
    for (const note of inactiveNotes) {
      activeNotes.delete(note);
    }
  }
  return Array.from(activeNotes);
```

shorter & cleaner imho

Reply:
slick 🤌